content/docs/rfcs/32/README.md

```diff
@@ -0,0 +1,560 @@
+Upon registration, they should receive `leaf_index` value which represents their position in the merkle tree. 
+Receiving a `leaf_index` is not a hard requirement and is application specific. 
+The other way around is the users calculating the `leaf_index` themselves upon successful registration.
+
+## Signaling
+
+After registration, 
+the users can participate in the application by sending signals to the other participants in a decentralised manner or to a centralised server. 
+Along with their signal, 
+they need to generate a ZK-Proof by using the circuit with the specification described above. 
+
+For generating a proof, 
+the users need to obtain the required parameters or compute them themselves, 
+depending on the application implementation and client libraries supported by the application.
+For example the users can store the membership merkle tree on their end and 
+generate a merkle proof whenever they want to generate a signal. 
+
+### Implementation notes
+
+#### Signal hash
+
+The signal hash can be generated by hashing the raw signal (or content) using the `keccak256` hash function.
+
+#### Epoch
+
+The epoch can be generated by hashing a raw string (i.e UNIX timestamp) value using `keccak256`.
+
+####  Obtaining merkle proof
+
+The merkle proof should be obtained locally or from a trusted third party. 
+By using the [incremental merkle tree algorithm](https://github.com/appliedzkp/incrementalquintree/blob/master/ts/IncrementalQuinTree.ts), 
+the merkle can be obtained by providing the `leaf_index` of the `identity_commitment`. 
+The proof (`merkle_proof`) is composed of the following fields:
+
+```
+{
+  root: bigint
+  indices: number[]
+  path_elements: bigint[][]
+}
+```
+
+
+#### Generating proof
+
+For proof generation, 
+the user need to submit the following fields to the circuit:
+
+```
+    {
+      identity_secret: identity_secret_hash,
+      path_elements: merkle_proof.path_elements,
+      identity_path_index: merkle_proof.indices,
+      x: signal_hash,
+      epoch: epoch,
+      rln_identifier: rln_identifier
+    }
+```
+
+
+#### Calculating output
+
+The proof output is calculated locally, 
+in order for the required fields for proof verification to be sent along with the proof. 
+The proof output is composed of the `y` share of the secret equation and the `internal_nullifier`. 
+The following fields are needed for proof output calculation:
+
+```
+{
+  identity_secret_hash: bigint, 
+  epoch: bigint, 
+  rln_identifier: bigint,
+  x: bigint, 
+}
+```
+
+The output `[y, internal_nullifier]` is calculated in the following way:
+
+```
+a_0 = identity_secret
+a_1 = poseidonHash([a0, epoch])
+
+y = a_0 + x * a_1
+
+internal_nullifier = poseidonHash([a_1, rln_identifier])
+```
+
+
+#### Sending the output message
+
+The user's output message (`output_message`), 
+containing the signal should contain the following fields at minimum:
+
+```
+{
+    content: signal, # non-hashed signal
```

Comment:
wondering why the key-value pair format? 
e.g., `signal` is already a named variable, it is a bit confusing to use the `content` term for it.  Can `content` contain anything rather than `signal`?

Reply:
You're right, let's just name it signal